lib/verbose/output.js

```diff
@@ -0,0 +1,68 @@
+import {inspect} from 'node:util';
+import stripFinalNewline from 'strip-final-newline';
+import {escapeLines} from '../arguments/escape.js';
+import {PIPED_STDIO_VALUES} from '../stdio/forward.js';
+import {verboseLog} from './log.js';
+
+export const handleStreamsVerbose = ({stdioStreams, options, isSync, stdioState, verboseInfo}) => {
+	if (!shouldLogOutput(stdioStreams, options, isSync, verboseInfo)) {
+		return stdioStreams;
+	}
+
+	const [{fdNumber}] = stdioStreams;
+	return [...stdioStreams, {
+		...stdioStreams[0],
+		type: 'generator',
+		value: verboseGenerator.bind(undefined, {stdioState, fdNumber, verboseInfo}),
+	}];
+};
+
+// `ignore` opts-out of `verbose` for a specific stream.
+// `ipc` cannot use piping.
+// `inherit` would result in double printing.
+// They can also lead to double printing when passing file descriptor integers or `process.std*`.
+// This only leaves with `pipe` and `overlapped`.
+const shouldLogOutput = (stdioStreams, {encoding}, isSync, {verbose}) => verbose === 'full'
+	&& !isSync
+	&& ALLOWED_ENCODINGS.has(encoding)
+	&& fdUsesVerbose(stdioStreams)
+	&& (stdioStreams.some(({type, value}) => type === 'native' && PIPED_STDIO_VALUES.has(value))
+	|| stdioStreams.every(({type}) => type === 'generator'));
+
+// Only print text output, not binary
+// eslint-disable-next-line unicorn/text-encoding-identifier-case
+const ALLOWED_ENCODINGS = new Set(['utf8', 'utf-8']);
+
+// Printing input streams would be confusing.
+// Files and streams can produce big outputs, which we don't want to print.
+// We could print `stdio[3+]` but it often is redirected to files and streams, with the same issue.
+// So we only print stdout and stderr.
+const fdUsesVerbose = ([{fdNumber}]) => fdNumber === 1 || fdNumber === 2;
```

Comment:
The comments in this file describe the reasons behind some of the restrictions for the `verbose: 'full'` option described in `readme.md`.